Smart-NIC Support Phase1

adrianchiris opened the pull request:
**- What this PR does and why is it needed**
This PR add support for Smart-NICs in OVN-Kubernetes.

It addresses Issue #1558 The included design doc provides additional context and information.

Support for Smart-NICs is Planned to be implemented in phases, each phase enables additional traffic flows.

Phase 1 (This PR) introduces the concept of smart-NIC and enables ovnkube node logic separation when running either on a regular host, a smart-NIC host or a smart-NIC embedded CPU.

Traffic flows which are supported for phase 1:
- Pod to Pod
- Pod to Service
- Pod to External

These are essentially the same changes as merged to the `bluefield` branch rebased on top of master and partitioned into logically manageable commits.

On Phase 2 we plan to support management port for smart-NICs to allow hostNetwork pods access other pods and services.



**- How to verify it**
Unit-tests are still WIP

**- Description for the changelog**
Smart-NIC support Phase-1

adrianchiris commented:
/cc @girishmg

moshe010 commented on dist/images/ovnkube.sh:
```diff
@@ -1030,14 +1059,15 @@ ovn-node() {
     --logfile-maxage=${ovnkube_logfile_maxage} \
     ${hybrid_overlay_flags} \
     ${disable_snat_multiple_gws_flag} \
+    --k8s-cacert=${K8S_CACERT} \
```
same question

adrianchiris replied:
When ovnkube node running on Smart-NIC, it requires K8S token and CA cert to access K8s API.
these need to be provided to ovnkube node.

looking more closely, we should provide both the same way, either through CLI or Env variables.
I will remove this so both will be taken from Environment variables for consistency.
(with current code, one is taken from ENV variable and the other from CLI flag)

moshe010 commented on dist/images/ovnkube.sh:
```diff
@@ -1030,14 +1059,15 @@ ovn-node() {
     --nodeport \
     --mtu=${mtu} \
-    ${OVN_ENCAP_IP} \
+    ${ovn_encap_ip_flag} \
```
why this change?

adrianchiris replied:
This exposes a way to explicitly set OVN encapsulation IP.

A first time run of ovnkube node will derrive this IP address from `util.GetNodePrimaryIP()` which takes the IP address of the k8s node object.

In case ovnkube node runs on Smart-NIC,  the OVN encap IP != Node IP

moshe010 replied:
thank make sense

moshe010 commented on go-controller/go.mod:
```diff
@@ -3,7 +3,7 @@ module github.com/ovn-org/ovn-kubernetes/go-controller
 
 require (
-	github.com/Mellanox/sriovnet v1.0.1
+	github.com/Mellanox/sriovnet v1.0.2-0.20210112115703-ecc40df73c7c
```
I released github.com/Mellanox/sriovnet v1.0.2

adrianchiris replied:
Will update to the released version, thanks!

moshe010 commented on go-controller/pkg/cni/cniserver.go:
```diff
@@ -46,7 +48,11 @@ import (
 // NewCNIServer creates and returns a new Server object which will listen on a socket in the given path
-func NewCNIServer(rundir string, factory factory.NodeWatchFactory) *Server {
+func NewCNIServer(rundir string, factory factory.NodeWatchFactory, kclient kubernetes.Interface, mode string) (*Server, error) {
+	if mode == types.NodeModeSmartNIC {
```
you will never hit this code? do we need this check ...

see 
	if config.OvnKubeNode.Mode == types.NodeModeSmartNIC {
	err = cniServer.Start(cni.HandleCNIRequest)			n.watchSmartNicPods()
	} else {
		// start the cni server
		kclient, ok := n.Kube.(*kube.Kube)
		if !ok {
			return fmt.Errorf("cannot get kubeclient for starting CNI server")
		}
		var cniServer *cni.Server
		cniServer, err = cni.NewCNIServer("", n.watchFactory, kclient.KClient, config.OvnKubeNode.Mode)
		if err == nil {
			err = cniServer.Start(cni.HandleCNIRequest)
		}
	}

adrianchiris replied:
Yes i am aware.

Should we just rely on the consumer of cniserver to provide valid mode ? 
or should cniserver also validate (as if we got here, someone did something unexpected and we should fail)?
in the latter case perhaps i should make the validation more strict (as only "full" and "smar-nic-host" are supported in case additional modes are added down the road.)

moshe010 replied:
I see, I missed the part that you need the mode for the NewPodLister

adrianchiris replied:
Originally i didnt want the `NewCNIServer()` method to rely on config for object creation. seemed to me like the caller will have a better understands towards what he is going to get from `NewCNIServer()` if this parameter is passed to this function.

However if you feel its simpler to just use config directly im fine with changing it.

moshe010 commented on go-controller/pkg/cni/helper_linux.go:
```diff
@@ -275,29 +250,9 @@ func setupSriovInterface(netns ns.NetNS, containerID, ifName string, ifInfo *Pod
-		return nil, err
-	}
-
+func ConfigureOVS(ctx context.Context, namespace string, podName string, hostIfaceName string,
```
add method description to  ConfigureOVS

moshe010 commented on go-controller/pkg/cni/helper_linux.go:
```diff
@@ -309,41 +264,79 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInf
+		l, err := netlink.LinkByName(hostIfaceName)
 		if err != nil {
-			return nil, fmt.Errorf("failed to find host veth interface %s: %v", hostIface.Name, err)
+			return fmt.Errorf("failed to find host veth interface %s: %v", hostIfaceName, err)
```
not related to the pr but we should s/veth interface/interface

adrianchiris replied:
as discuss in the other comment about this topic. it should be handled in a different PR

moshe010 commented on go-controller/pkg/cni/helper_linux.go:
```diff
@@ -309,41 +264,79 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInf
 		err = netlink.LinkSetTxQLen(l, 1000)
 		if err != nil {
-			return nil, fmt.Errorf("failed to set host veth txqlen: %v", err)
+			return fmt.Errorf("failed to set host veth txqlen: %v", err)
```
s/veth/interface

adrianchiris replied:
not directly related to this PR IMO.
(although tis is a bit missleading with SR-IOV switchdev)

moshe010 replied:
Let fix it in different PR

moshe010 commented on go-controller/pkg/cni/utils.go:
```diff
@@ -0,0 +1,80 @@
+
+// wait cond smart-NIC: wait for OVN master to set pod-networks annotation and
+// ovnkube running on Smart-NIC to set connection-status pod annotation and its status is Ready
+func smartNICReady(podAnnotation map[string]string) bool {
```
maybe isSmartNICReady

moshe010 commented on go-controller/pkg/cni/utils.go:
```diff
@@ -0,0 +1,80 @@
+type podAnnotWaitCond func(podAnnotation map[string]string) bool
+
+// wait cond for OVN master to set pod-networks annotation
+func ovnReady(podAnnotation map[string]string) bool {
```
maybe isOvnReady

moshe010 commented on go-controller/pkg/config/config.go:
```diff
@@ -1637,3 +1671,36 @@ func UpdateOVNNodeAuth(masterIP []string, southboundDBPort, northboundDBPort str
+		return err
+	}
+
+	// validate ovnkube-node-mode
```
can we warp this validation in a function?

adrianchiris replied:
Sure

moshe010 commented on go-controller/pkg/node/node.go:
```diff
@@ -201,40 +208,43 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
+		if err := nodeAnnotator.Run(); err != nil {
+			return fmt.Errorf("failed to set node %s annotations: %v", n.name, err)
+		}
+		go n.gateway.Run(n.stopChan, wg)
```
you move go n.gateway.Run to here on propose? in the old code it first wait for management port

adrianchiris replied:
a mistake, it should be after `waiter` was invoked , good catch !

moshe010 commented on go-controller/pkg/node/node.go:
```diff
@@ -201,40 +208,43 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
-	klog.Infof("Gateway and management port readiness took %v", time.Since(start))
 
 	if config.HybridOverlay.Enabled {
+		// Not supported with Smart-NIC, enforced in config
```
should we add TODO and revisit this in the feature?

adrianchiris replied:
TODO added

moshe010 commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+			if !util.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				return
+			}
+			if util.PodScheduled(pod) {
```
do we still need this check I thought we listen on pod on the same node?

adrianchiris replied:
it actually depends on the type of watch factory (master or node),  only the node watch factory limits listening on pods on a specific node.

looking at `ovnkube.go` there is a mode where both master and node run in the same process, in this case the the master watch factory will be used which does not limit the scope to a specific node.

https://github.com/ovn-org/ovn-kubernetes/blob/0a2aad51333400e28fbf51c1017b3064d61186cc/go-controller/cmd/ovnkube/ovnkube.go#L266


IMO it is safer to not assume filtration is performed at informer level.

moshe010 replied:
Ok make sense

moshe010 commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+				return
+			}
+			_, retry := retryPods.Load(pod.UID)
+			if util.PodScheduled(pod) && retry {
```
same question on the util.PodScheduled(pod)

adrianchiris replied:
Please see comment above.

moshe010 commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to setup representor port. failed to set pod annotations. %v", err)
+	}
+
```
maybe put Line 163-167 in defer it duplicate in171-176

adrianchiris replied:
I originally considered that, however i didnt see much use of defer bering leveraged for (non trivial) cleanup  in non test packages of the project.

doing it proper we should defer multiple methods  i.e:
at L137 - `defer func() {if err != nil {n.delRepPort()}}`
at L157 - `defer func() {if err != nil {util.GetNetLinkOps.LinkSetDown(link)}}`

adrianchiris replied:
sticking to original impl here for now.

moshe010 reviewed (approved):
LGTM

Billy99 reviewed (approved):
Minor nits.

I am in the process of trying to get this PR install on our local BF deployment. I'll let you know if I find anything.

zshi-redhat commented on go-controller/pkg/cni/helper_linux.go:
```diff
@@ -309,41 +265,79 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInf
+		// SR-IOV Case
+		hostIface, contIface, err = setupSriovInterface(netns, pr.SandboxID, pr.IfName, ifInfo, pr.CNIConf.DeviceID)
+	} else {
+		// General case
```
Shall we return immediately when `pr.CNIConf.DeviceID == "" && pr.IsSmartNIC == false` ? Because there is no OVS running in host mode, we know it would fail.

adrianchiris replied:
Agree, better to catch this here.

zshi-redhat commented on dist/templates/ovnkube-node.yaml.j2:
```diff
@@ -259,6 +259,13 @@ spec:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: nvidia.com/smart-nic
```
Do we want to define a general label using ovn-k8s prefix?

adrianchiris replied:
yes we do. this is actually in my TODO list.

i will update the prefix to `k8s.ovn.org`

zshi-redhat commented on dist/images/ovnkube.sh:
```diff
@@ -1046,10 +1075,13 @@ ovn-node() {
-    --metrics-bind-address ${ovnkube_node_metrics_bind_address} &
+    --metrics-bind-address ${ovnkube_node_metrics_bind_address} \
+     ${ovnkube_node_mode_flag} &
 
```
Shall we remove these options in smart-nic-host mode? or maybe add conditional check when generating these options above (line 1035).

ovn_node_ssl_opts
metrics-bind-address

adrianchiris replied:
> ovn_node_ssl_opts

yes we should not expose them for ovnkube-node in smart-nic-host mode.

> metrics-bind-address

So we are technically still starting an ovnkube node. we dont need to report metrics on it ?

https://github.com/ovn-org/ovn-kubernetes/blob/7f932cac4e69232c85483e06cd6b9772914fe04d/go-controller/cmd/ovnkube/ovnkube.go#L285

i think we would want metrics on ovnkube Node in all modes.

zshi-redhat commented:
Run some tests with this PR applied:
1) image can be built for both x86 and arm arches successfully
2) ovnkube-node runs in smart-nic-host mode on openshift worker
3) ovnkube-node runs in smart-nic mode on BF, ovn-controller runs on BF
4) ovnkube-node runs in full mode in non-BF workers
5) VF and rep can be attached to sriov pod on the host and br-int bridge on BF separately
6) sriov pod to regular pod traffic is working, performance is ~3Gbits/s.

Looking at why performance is not good. 
1) tcpdump on vf rep shows no packets packets during pod-to-pod iperf runs which indicates traffic is offloaded to hardware
2) tc filter counter shows hardware packet counter is increasing
3) will test with two sriov pods.

adrianchiris commented on go-controller/pkg/cni/types.go:
```diff
@@ -22,9 +24,10 @@ const serverSocketPath string = serverRunDir + "/" + serverSocketName
+	MTU        int   `json:"mtu"`
+	Ingress    int64 `json:"ingress"`
+	Egress     int64 `json:"egress"`
+	IsSmartNic bool  `josn:"smartnic"`
```
Good catch ! will fix that.

trozet replied:
this is still wrong

adrianchiris commented on go-controller/pkg/cni/utils.go:
```diff
@@ -0,0 +1,81 @@
+	return false
+}
+
+// getPodAnnotations obtains the pod annotation from the cache
```
will be adressed.

trozet replied:
the doctext is still not right here

adrianchiris commented on dist/templates/ovnkube-node-smart-nic-host.yaml.j2:
```diff
@@ -0,0 +1,187 @@
+---
```
I agree with you.
Let me give it a go, will address this as an additional commit in this PR

adrianchiris commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to get link device for interface %s", vfRepName)
+	}
+
```
im thinking we should have the same behaviour as hw-offload VF in non smart-nic case. 
so we should rename it.

im not sure if this PR should address restoring the VF rep name to its original name (i.e implement some local cache)
as we dont restore VF rep name in non Smart-NIC case.

zshi-redhat replied:
Or maybe we don't want to rename the VF rep, using its original name would be easy for debugging.

If renaming is preferred, I think we can address the restore in another PR.

adrianchiris replied:
@moshe010 do you have any preference here ?

adrianchiris replied:
So Per PR #2066 i will leave VF rep name as is for now.

we can align this bit later for hw offload VF if its decided to go twards a different direction.

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			klog.Infof("DeleteFunc for POD: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			if _, ok := servedPods.Load(pod.UID); !ok {
```
Do you really need servedPods map? What is the harm in just trying to delete no matter what?

adrianchiris replied:
missed that yesterday when addressing comments.
So we used the servedPods map to avoid:
- Possible error log when failing to get VF representor from pod annotation
- unneeded netdev operations

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+
+// addRepPort adds the representor of the VF to the ovs bridge
+func (n *OvnNode) addRepPort(pod *kapi.Pod, vfRepName string, ifInfo *cni.PodInterfaceInfo) error {
+	klog.Infof("addRepPort: %s", vfRepName)
```
lets not use the function name in the log message. Change it to something more obvious like "Adding VF representor %s for pod %s"

adrianchiris replied:
will be addressed

trozet replied:
need to fix in this commit

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+		return fmt.Errorf("failed to get link device for interface %s", vfRepName)
+	}
+
+	klog.Infof("addRepPort: set link mtu %s", vfRepName)
```
this and the link up logs don't seem very helpful. Just log the error returned by this function.

adrianchiris replied:
removed

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+// delRepPort delete the representor of the VF from the ovs bridge
+func (n *OvnNode) delRepPort(vfRepName string) error {
+	//TODO(adrianc): handle: clearPodBandwidth(pr.SandboxID), pr.deletePodConntrack()
+	klog.Infof("delRepPort: %s", vfRepName)
```
make  a clearer log message like mentioned above dont use the function name

adrianchiris replied:
will be addressed

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+	//TODO(adrianc): handle: clearPodBandwidth(pr.SandboxID), pr.deletePodConntrack()
+	klog.Infof("delRepPort: %s", vfRepName)
+	// Set link down for representor port
+	link, err := util.GetNetLinkOps().LinkByName(vfRepName)
```
why do you need to set it down before you delete it?

adrianchiris replied:
we can also do it after IMO. i will swap the order as first priority is to get it removed from OVS.

trozet commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+			klog.Warningf("Failed to set link down for representor port %s. %v", vfRepName, linkDownErr)
+		}
+	}
+	klog.Infof("Port %s link state set to \"down\"", vfRepName)
```
this should be a debug message

adrianchiris replied:
i will remove this. i dont think it adds much value

trozet commented on dist/templates/ovnkube-node-smart-nic-host.yaml.j2:
```diff
@@ -0,0 +1,187 @@
+
+      nodeSelector:
+        kubernetes.io/os: "linux"
+        nvidia.com/smart-nic: ""
```
why nvidia? can we use a k8s.ovn.org label?

adrianchiris replied:
there is another comment about it, i will change to k8s.ovn.org prefix.

trozet reviewed (changes requested):
overall looks good. Left some comments inline. Please also add missing unit test. The coverage is reporting 1% drop.

adrianchiris commented on go-controller/pkg/cni/cniserver.go:
```diff
@@ -222,13 +230,16 @@ func (s *Server) handleCNIRequest(r *http.Request) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	if s.mode == types.NodeModeSmartNICHost {
```
I wanted to avoid reaching into the globally defined structs in config package if possible as imo its a bit clearer if cniserver bases its decisions on the values of internal members instead of globally defined values.

trozet replied:
OK up to you, but in my experience we typically allow config var direct access across the project

adrianchiris commented on go-controller/pkg/cni/cnismartnic.go:
```diff
@@ -0,0 +1,54 @@
+package cni
+
+import (
+	"fmt"
```
done.

trozet replied:
not fixed in this commit

adrianchiris commented on go-controller/pkg/cni/cnismartnic.go:
```diff
@@ -0,0 +1,54 @@
+		return err
+	}
+
+	// 4. Set smart-nic connection-details pod annotation
```
Done.

trozet replied:
not fixed

adrianchiris commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+			}
+			if util.PodScheduled(pod) {
+				// Is this pod created on same node where the smart NIC
+				if n.name != pod.Spec.NodeName {
```
we are using filtered informer in ovnkube, we just dont rely that.

adrianchiris commented on go-controller/pkg/node/node_smartnic.go:
```diff
@@ -0,0 +1,203 @@
+				vfRepName, err := n.getVfRepName(pod)
+				if err != nil {
+					klog.Infof("Failed to get rep name, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
```
Yes, this is needed just for the first time.
Looking at the implementation of `Store()`, it will re-assign the value instead of skip as it operates on pointers.

adrianchiris replied:
removed my original reply here,

looking at this agian, i believe you are right as L#69 ensures that

trozet replied:
but it's still in this commit

adrianchiris commented:
@trozet @Billy99 @zshi-redhat i have addressed your comments.
Did that in an additional Commit to make reviewing the change easier.

I will start working on unit tests.

adrianchiris commented:
> I will start working on unit tests.

Would it be possible to do that in a separate PR ?

Will also rebase as i see there are conflicts.

zshi-redhat commented:
> @trozet @Billy99 @zshi-redhat i have addressed your comments.
> Did that in an additional Commit to make reviewing the change easier.

@adrianchiris the new change looks good to me.
> 
> I will start working on unit tests.

moshe010 commented:
ovnkube write the cni config in our case (smart-nic host) we can skip it. it just causing issues https://github.com/ovn-org/ovn-kubernetes/blob/2d2424295599626b644de0e0c52f86f993e8056c/go-controller/pkg/node/node.go#L266-L271

zshi-redhat commented:
> ovnkube write the cni config in our case (smart-nic host) we can skip it. it just causing issues
> 
> https://github.com/ovn-org/ovn-kubernetes/blob/2d2424295599626b644de0e0c52f86f993e8056c/go-controller/pkg/node/node.go#L266-L271

What is the issue? The cni config is needed for smart-nic-host mode, isn't it?

moshe010 commented:
> > ovnkube write the cni config in our case (smart-nic host) we can skip it. it just causing issues
> > https://github.com/ovn-org/ovn-kubernetes/blob/2d2424295599626b644de0e0c52f86f993e8056c/go-controller/pkg/node/node.go#L266-L271
> 
> What is the issue? The cni config is needed for smart-nic-host mode, isn't it?

the configure of ovn-k8s-cni-overlay is in the net-attach-def here we just used muluts. but if existing of /etc/cni/net.d/10-ovn-kubernetes.conf will cause issues when you restart multus as it will put it in the multus delegate part. Anwy we don't use /etc/cni/net.d/10-ovn-kubernetes.conf so we don't need to create it

adrianchiris commented:
@moshe010 @zshi-redhat 

I have reverted my last commit which prevents ovnkube node in smart-nic-host mode from creating CNI config.

Reason for revert:
1. not creating the CNI config will make deployment harder as multus CNI will need to be deployed with special config as the "auto" config mode will fail on smart-nic nodes.
2. deploying ovn-kubernetes without multus, as no CNI is generated at host level will cause the node to report as not-ready.
    this is undesirable from cluster depoyment perspective (e.g in OpenShift it will cause issues)

zshi-redhat commented:
> @moshe010 @zshi-redhat
> 
> I have reverted my last commit which prevents ovnkube node in smart-nic-host mode from creating CNI config.
> 
> Reason for revert:
> 
>     1. not creating the CNI config will make deployment harder as multus CNI will need to be deployed with special config as the "auto" config mode will fail on smart-nic nodes.
> 
>     2. deploying ovn-kubernetes without multus, as no CNI is generated at host level will cause the node to report as not-ready.
>        this is undesirable from cluster depoyment perspective (e.g in OpenShift it will cause issues)

Thanks @adrianchiris !

zshi-redhat commented:
> > @moshe010 @zshi-redhat
> > I have reverted my last commit which prevents ovnkube node in smart-nic-host mode from creating CNI config.
> > Reason for revert:
> > ```
> > 1. not creating the CNI config will make deployment harder as multus CNI will need to be deployed with special config as the "auto" config mode will fail on smart-nic nodes.
> > 
> > 2. deploying ovn-kubernetes without multus, as no CNI is generated at host level will cause the node to report as not-ready.
> >    this is undesirable from cluster depoyment perspective (e.g in OpenShift it will cause issues)
> > ```
> 
> Thanks @adrianchiris !

Does it make sense to allow overwriting CNI config file (e.g. adding an argument for ovnkbue called -cni-conf-file)?
User or other component like SR-IOV Operator can define the new default network by adding a resource annotation or clusterNetwork in the cni-conf-file.

trozet commented:
@adrianchiris can you please rebase?

trozet commented:
@adrianchiris still have some conflicts here, sorry :( can you rebase again please?

adrianchiris commented:
@trozet 

Rebased against master and added unit tests.
I have tested this on a Smart-NIC deployment, pod to pod and pod to service traffic is working

Main pain-points for the rebase was in `pkg/node/node.go::OvnNode.Start()`

Note:
The traffic is not offloaded, however we see the same with master branch running on HW offload NICs, so i dont believe its related to this PR.

trozet commented:
@adrianchiris error: error parsing ovnkube-node.yaml: error converting YAML to JSON: yaml: line 83: mapping values are not allowed in this context

adrianchiris commented:
looking at the logs : `Error: Bad response: 504 ` when trying to submit code coverage results to coveralls

adrianchiris commented:
/retest

adrianchiris commented:
@trozet appreciate a look at the updated PR, i dont think there are outstanding comments.

the failed e2e test does not seem to be related to this PR as it fails on kubectl timeout

trozet commented:
hmm looks like a flake with db delete:
Mar 30 14:32:17.273: INFO: connectivity test before deleting db files

trozet commented on go-controller/go.mod:
```diff
@@ -70,5 +70,4 @@ replace (
 	k8s.io/metrics => k8s.io/metrics v0.20.0
 	k8s.io/mount-utils => k8s.io/mount-utils v0.20.0
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.20.0
-
```
this commit doesn't look like its doing anything anymore? should we remove it?

adrianchiris replied:
good catch, i see sriovnet was version bumped in https://github.com/ovn-org/ovn-kubernetes/commit/bfe617c45318a6a0b8d80bc3ae6823cf63c1585d

will drop this commit

trozet commented on go-controller/pkg/util/smartnic_annotations.go:
```diff
@@ -0,0 +1,128 @@
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
```
nit: separate this with a new line, its a project package import

adrianchiris replied:
will do.

trozet reviewed (changes requested):
thanks for adding the unit tests! still some various minor things in commits that haven't been fixed. PTAL.

adrianchiris commented:
@trozet it seems like the old comments were addressed in the wrong commit :(
https://github.com/ovn-org/ovn-kubernetes/pull/2005/commits/195ccfec41a14785742b36c5c38cf49bf237854b
probably due to a rebase that went wrong on my part. will move to the relevant commit.
